examples/agents/ppo-chess-llm.py

```diff
@@ -0,0 +1,478 @@
+
+    class CriticHead(torch.nn.Module):
+        def __init__(self):
+            super().__init__()
+            self.m = torch.nn.Linear(3584, 1, dtype=torch.bfloat16)
+
+        def forward(self, hidden):
+            return self.m(hidden).squeeze(-1)
+
+    critic_llm_policy = Seq(
+        Mod(CriticHead(), in_keys=["hidden"], out_keys=["state_value"]),
+    )
+
+    return actor_llm_policy, data_llm_policy, critic_llm_policy, tokenizer
+
+
+def play(env, data_llm_policy, actor_llm_policy, tokenizer):
+    from torchrl.data import LazyStackStorage, ReplayBuffer
+
+    rb = ReplayBuffer(
+        storage=LazyStackStorage(100),
+        batch_size=32,
+        sampler=SliceSamplerWithoutReplacement(slice_len=8, end_key=("next", "done")),
+    )
+    # layout=torch.jagged errors with Qwen
+    # File "/home/mg1998/.conda/envs/rl/lib/python3.10/site-packages/transformers/models/qwen2/modeling_qwen2.py", line 859, in forward
+    # cache_position = torch.arange(
+    #             past_seen_tokens, past_seen_tokens + inputs_embeds.shape[1], device=inputs_embeds.device
+    #         )
+    # AttributeError: 'ConstantIntNode' object has no attribute 'add'
+    # After attempting to fix this there were other issues like NJT can't unsqueeze dim=0
+    # rb.append_transform(lambda td: td.densify(layout=torch.jagged))
+    # rb.append_transform(
+    #     lambda td: td.set(
+    #         "obs_tokens", td.get("obs_tokens").to_padded_tensor(tokenizer.pad_token_id)
+    #     )
+    # )
+
+    # rb.append_transform(lambda td: td.set("tokenized_action", td.get("tokenized_action").to_padded_tensor(tokenizer.pad_token_id)))
+
+    collector = SyncDataCollector(
+        env, data_llm_policy, frames_per_batch=100, total_frames=10000
+    )
+    loss_module = ClipPPOLoss(
+        actor_network=actor_llm_policy,
+        critic_network=critic_llm_policy,
+    )
+
+    # FIXME: is this the right way to do this?
+    loss_module.tensor_keys.action = "tokenized_action"
+
+    optim = torch.optim.Adam(loss_module.parameters())
+
+    gae = GAE(
+        value_network=Seq(actor_llm_policy[0], *critic_llm_policy),
+        gamma=0.99,
+        lmbda=0.95,
+        shifted=True,
+    )
+
+    import torch.nn.functional as F
+
+    def pad_tensors_to_same_shape(tensor1, tensor2):
+        size1 = tensor1.shape
+        size2 = tensor2.shape
+
+        if size1[-1] < size2[-1]:
+            pad1 = size2[-1] - size1[-1]
+            tensor1_padded = F.pad(tensor1, (0, pad1))
+            tensor2_padded = tensor2
+        elif size1[-1] > size2[-1]:
+            pad2 = size1[-1] - size2[-1]
+            tensor2_padded = F.pad(tensor2, (0, pad2))
+            tensor1_padded = tensor1
+        else:
+            tensor1_padded = tensor1
+            tensor2_padded = tensor2
+
+        return tensor1_padded, tensor2_padded
+
+    for data in tqdm(collector):
+        # FIXME: reward seems to be getting wrongly propagated (e.g. sunfish's win gets reflected as llm's win)
```

Comment:
debugging this one, not sure whether it's an issue with how I applied the transforms